p2p/peer.py

```diff
@@ -598,8 +598,17 @@ def __len__(self):
     def is_full(self) -> bool:
         return len(self) >= self.max_peers
 
+    def is_valid_connection_candidate(self, candidate: Node) -> bool:
+        # connect to no more then 2 nodes with the same IP
+        nodes_by_ip = groupby(
+            operator.attrgetter('remote.address.ip'), self.connected_nodes.values())
+        matching_ip_nodes = nodes_by_ip.get(candidate.address.ip, [])
+        return len(matching_ip_nodes) <= 2
+
     def get_nodes_to_connect(self) -> Generator[Node, None, None]:
```

Comment:
@williambannas not really related to this PR but as you already touch the bulk of this function you could change the return type to `Iterable[Node]` which seems more idiomatic for this use case